File: vito/litho/src/main/java/com/facebook/fresco/vito/litho/AboveViewportWorkingRange.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.litho;

import com.facebook.litho.WorkingRange;

/**
 * A working range targeting components that are above the viewport (i.e., components the user has
 * scrolled past).
 */
public class AboveViewportWorkingRange implements WorkingRange {
  private boolean isInRange(int position, int firstVisibleIndex) {
    return position < firstVisibleIndex;
  }

  @Override
  public boolean shouldEnterRange(
      int position,
      int firstVisibleIndex,
      int lastVisibleIndex,
      int firstFullyVisibleIndex,
      int lastFullyVisibleIndex) {
    return isInRange(position, firstVisibleIndex);
  }

  @Override
  public boolean shouldExitRange(
      int position,
      int firstVisibleIndex,
      int lastVisibleIndex,
      int firstFullyVisibleIndex,
      int lastFullyVisibleIndex) {
    return !isInRange(position, firstVisibleIndex);
  }
}

File: vito/litho/src/main/java/com/facebook/fresco/vito/litho/AboveViewportWorkingRange.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.litho

import com.facebook.litho.WorkingRange

/**
 * A working range targeting components that are above the viewport (i.e., components the user has
 * scrolled past).
 */
class AboveViewportWorkingRange : WorkingRange {

  private fun isInRange(position: Int, firstVisibleIndex: Int): Boolean =
      position < firstVisibleIndex

  override fun shouldEnterRange(
      position: Int,
      firstVisibleIndex: Int,
      lastVisibleIndex: Int,
      firstFullyVisibleIndex: Int,
      lastFullyVisibleIndex: Int
  ): Boolean = isInRange(position, firstVisibleIndex)

  override fun shouldExitRange(
      position: Int,
      firstVisibleIndex: Int,
      lastVisibleIndex: Int,
      firstFullyVisibleIndex: Int,
      lastFullyVisibleIndex: Int
  ): Boolean = !isInRange(position, firstVisibleIndex)
}
